### PR TITLE
Added resource for SQL MI "azurerm_sql_managed_instance"

### LINKED ISSUE #1747: Managed SQL database instance
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

---

It would be nice if we have a terraform feature for Azure SQL Managed database instance.

### PR DESCRIPTION
**Pull request by joranm**
New resource provider for SQL Managed Instance
https://docs.microsoft.com/en-us/azure/sql-database/sql-database-managed-instance

**Known issues**
- Destroys fails due to platform lease on Virtual Server Object (12 hours before it is deleted) 

Fixes #1747

### CODE REVIEW

**Comment by katbyte** [comment]
@joranm,

gometalinter removed some linters (megacheck and unused) as they were merged into staticcheck and we didn't pin the tooling version. As a result all linting was failing. It has been resolved in #2722 so merge/rebase on master and you should be g2g 🙂

**Comment by joranm** [comment]
Hi @katbyte, @tombuildsstuff , did a rebase, and tests have passed successfully.
Regarding the known issue. What would be the best approach?

**1) Timeout issue on create (6 hours):** Is it possible to implement the Timeout on these resources? I didn't found this feature implemented at any other resource.
**2) Lock on dependend resources** The creation of SQL Managed Instance does also create a Virtual Server Object  and a Network Intent Policy, which locks the subnet and routetable. This object is automatically cleaned after 12 hours after the SQL MI instance removal.

**Comment by katbyte** [comment]
@joranm,

wrt 1 - at the moment we don't have timeouts implemented, that is something we plan to implement for v2.0 so until then tested this will be.. difficult

**Comment by archmangler** [comment]
There is a suggestion on https://github.com/terraform-providers/terraform-provider-azurerm/issues/1747 that the creation time has been greatly reduced on the AzureRM side. Has anyone tested if this improves the situation sufficiently to make this provider usable?

**Comment by pkaminski-degree** [comment]
Is this feature planned to be merged any time soon?

**Comment by jacobm3** [comment]
Any idea when this resource will be merged? Would love to see this available. Thanks!

**Comment by sarayho1** [comment]
Will this resource be merged and released anytime soon?  Received a time-out error running the code with ARM JSON.

**Comment by TraGicCode** [comment]
Would be awesome to see more progress on getting this pushed through!

**Comment by WascawyWabbit** [comment]
Any news on when this will be merged?

**Comment by katbyte** [comment]
@joranm,

I went to run these tests today and got the following build error (as we have updated the azure go sdk)

```
==> Checking that code complies with gofmt requirements...
go install
# github.com/terraform-providers/terraform-provider-azurerm/azurerm
azurerm/resource_arm_sql_managed_instance.go:143:4: cannot use utils.String(licenseType) (type *string) as type "github.com/Azure/azure-sdk-for-go/services/preview/sql/mgmt/2015-05-01-preview/sql".ManagedInstanceLicenseType in field value
make: *** [build] Error 2
```

Went to merge master and push but it seems I can't? any chance you could update this se we could see if the instance provisions any faster now

**Comment by kim0** [comment]
I had opened #3593 and #3587 but apparently this is still WIP. If you have any questions, let me know though

**Comment by katbyte** [comment]
hi @kim0, the new PR is #3575 as this one seems to have been abandoned. Currently it is blocked with: 

```
------- Stdout: -------
=== RUN   TestAccAzureRMSqlMiServer_basic
=== PAUSE TestAccAzureRMSqlMiServer_basic
=== CONT  TestAccAzureRMSqlMiServer_basic
--- FAIL: TestAccAzureRMSqlMiServer_basic (741.07s)
    testing.go:568: Step 0 error: errors during apply:
        
        Error: Code="VnetSubnetConflictWithIntendedPolicy" Message="Route Table. (https://go.microsoft.com/fwlink/?linkid=871071)"
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test102924249/main.tf line 40:
          (source code not available)
        
        
FAIL
```

so any help on how to resolve that would be fantastic as i am working on other things at the moment.

**Comment by joranm** [comment]
> @joranm,
> I went to run these tests today and got the following build error (as we have updated the azure go sdk)
> ==> Checking that code complies with gofmt requirements...
> go install
> # github.com/terraform-providers/terraform-provider-azurerm/azurerm
> azurerm/resource_arm_sql_managed_instance.go:143:4: cannot use utils.String(licenseType) (type *string) as type "github.com/Azure/azure-sdk-for-go/services/preview/sql/mgmt/2015-05-01-preview/sql".ManagedInstanceLicenseType in field value
> make: *** [build] Error 2
> 
> Went to merge master and push but it seems I can't? any chance you could update this se we could see if the instance provisions any faster now

Hi @katbyte,
Provisioning of the first SQL MI in a vnet still takes between 6 and 12 hours. A lot of plumbing is done at the Azure side with regard to building a 5 node cluster, sql data gateway, loadbalancers and so on. I don't expect that the provisioning will be faster in the near future. Is it possible to implement the "timeout" property for Azure resources?

**Comment by katbyte** [comment]
@joranm, I believe that is in the works for v2.0 of the provider

**Comment by katbyte** [comment]
If you could pull in the changes from the PR i opened and or grant us push access to this branch we can take it from here

**Comment by tombuildsstuff** [comment]
> Provisioning of the first SQL MI in a vnet still takes between 6 and 12 hours.

@joranm 

Indeed, it's better than it was (36h) but it's still not ideal; chatting with the SQL Team a while back it turns out there's multiple conditions where the Create/Update call could take this long:

* when the cluster this is using under the hood within the subnet is initially built (as you're seeing)
* when the cluster is scaled up (for every 5 or so instances/databases within the same subnet)
* when the cluster is scaled down (for every 5 or so instances/databases within the same subnet)

> Is it possible to implement the "timeout" property for Azure resources?

As @katbyte has mentioned this is something we're doing as a part of the 2.0 release - at this point in time we're working on the dependencies to make that possible (we're working on switching out the Azure Storage SDK at the moment) at which point we should be able to enable Timeout support. However since this requires the Timeout support/this is going to ship in the 2.0 timeframe - I'm going to add this to the 2.0 milestone.

**Comment by kim0** [comment]
I'm using this provider (built from HEAD) quite a bit now. I'd like to also add other issues I'm see'ing

* Destroying a sql-mi cluster often fails. This is bec a "VirtualCluster" is deployed behind the scenes and that is not deletable .. at least not quickly. I talked to support, and they mentioned after all sql-mi instances are deleted, the VC is garbage collected (I think it was 7hours later). The act of destroying it also takes a couple of hours. TBH the whole SQLMI story on Azure feels very slow and not very cloud like sigh

When I hit the timeout deploying sqlmi .. terraform errors out. I wait ~ 4 hours. Then I run TF again, and it usually succeeds (bec sqlmi has finished deploying by then!).

**Comment by kim0** [comment]
Hello @katbyte .. Since the discussion seems to be happening here, [my comment](https://github.com/terraform-providers/terraform-provider-azurerm/pull/3575#issuecomment-498834050) contains the needed HCL to deploy without hitting the `VnetSubnetConflictWithIntendedPolicy` error.

Why is this removed from the 1.30 milestone. I mean it's already useful! Sure in some cases people will hit a timeout, but as in my previous comment just rerunning this again a few hours later completes successfully. So IMO it would be better to get this provider merged soon and only improve timeout handling in 2.0 ? As a user this would surely be better for me, than not being able to deploy sql-mi resource to begin with. I hope that makes sense

Thanks all

**Comment by tombuildsstuff** [comment]
@kim0 as mentioned above, unfortunately we're unable to ship this until this supports custom timeouts (which will be in 2.0) so that this can be provisioned in a single `terraform apply` - so whilst the approach you're describing may solve this for you, that's actually a separate bug [which is going to be fixed in 2.0](https://github.com/terraform-providers/terraform-provider-azurerm/issues/2807).

File: azurerm/resource_arm_sql_managed_instance.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+				Required:  true,
+				Sensitive: true,
+			},
+
+			"vcores": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  8,
+				ValidateFunc: validate.IntInSlice([]int{
+					8,
```
Comment (sheigel) [inline azurerm/resource_arm_sql_managed_instance.go]: A few weeks ago a smaller number of vcores (4) was introduced:
https://docs.microsoft.com/en-us/azure/sql-database/sql-database-managed-instance-resource-limits#instance-level-resource-limits

It seems worthwhile to have this new resource up to date.

**Comment by katbyte** [comment]
Hi @joranm, i'm going to close this as @mbfrahry has started to move this forward in #5399

**Pull request closed without merging**